engine/src/witness/sol/oracle_witnessing.rs

```diff
@@ -0,0 +1,383 @@
+// We use a prefunded account for this purpose. The keys have been burnt.
+#[allow(dead_code)]
+const PREFUNDED_ACCOUNT: SolAddress = const_address("CsS34ewTFLGqrpckPRww5hbWr4QJQ1J3ZA5D7WL4Ni3K");
+
+#[allow(dead_code)]
+pub struct PriceFeedData {
+	pub round_id: u32,
+	pub slot: u64,
+	pub timestamp: u32,
+	pub answer: i128,
+	pub decimals: u8,
+	pub description: String,
+}
+
+#[allow(dead_code)]
+pub async fn get_price_feeds<SolRetryRpcClient>(
+	sol_client: &SolRetryRpcClient,
+	oracle_query_helper: SolAddress,
+	oracle_program_id: SolAddress,
+	feed_addresses: Vec<SolAddress>,
+	min_context_slot: Option<u64>,
+) -> Result<Vec<PriceFeedData>, anyhow::Error>
+where
+	SolRetryRpcClient: SolRetryRpcApi + Send + Sync + Clone,
+{
+	let serialized_transaction = build_and_serialize_query_transaction(
+		oracle_query_helper,
+		oracle_program_id,
+		feed_addresses,
+	)
+	.map_err(|e| anyhow::anyhow!("Failed to build and serialize the query transaction: {:?}", e))?;
+
+	let simulation_result =
+		sol_client.simulate_transaction(serialized_transaction, min_context_slot).await;
+
+	let return_data = simulation_result
+		.value
+		.return_data
+		.as_ref()
+		.expect("Expected return data to be Some");
+
+	let price_feeds = decode_query_return_data(return_data, oracle_query_helper)
+		.map_err(|e| anyhow::anyhow!("Failed to decode the query return data: {:?}", e))?;
+
+	Ok(price_feeds)
+}
+
+// NOTE: This builds a transaction with the default compute units (200k). This should be enough for
+// querying more than 10 feeds so we don't bother extending the compute budget. If that was not
+// enough, the compute budget extension instruction needs to be added to the transaction before
+// serialization.
+fn build_and_serialize_query_transaction(
+	oracle_query_helper: SolAddress,
+	oracle_program_id: SolAddress,
+	feed_addresses: Vec<SolAddress>,
+) -> Result<Vec<u8>, anyhow::Error> {
+	let price_feed_metas: Vec<AccountMeta> = feed_addresses
+		.into_iter()
+		.map(|feed_account| AccountMeta::new(feed_account.into(), false))
+		.collect();
+	let instructions = vec![OracleQueryHelperProgram::with_id(oracle_query_helper)
+		.query_price_feeds(oracle_program_id)
+		.with_additional_accounts(price_feed_metas)];
+
+	let transaction = SolVersionedTransaction::new_unsigned(SolVersionedMessage::new(
+		&instructions,
+		Some(PREFUNDED_ACCOUNT.into()),
+		None,
+		&[],
+	));
+	transaction
+		.clone()
+		.finalize_and_serialize()
+		.map_err(|e| anyhow::anyhow!("Failed to serialize oracle query transaction: {:?}", e))
+}
+
+// Expected returned data is Vec<PriceFeedData>
+fn decode_query_return_data(
+	return_data: &UiTransactionReturnData,
+	expected_program_id: SolAddress,
+) -> Result<Vec<PriceFeedData>, anyhow::Error> {
+	let decoded_return_data = BASE64_STANDARD.decode(return_data.data.0.clone())?;
+	if return_data.data.1 != UiReturnDataEncoding::Base64 {
+		return Err(anyhow::anyhow!(
```

Comment:
```suggestion
		bail!(
```
same below - for early returns of an anyhow error, we prefer bail